feat: bouncer command to check how many validator witnessed something

marcellorigotti opened the pull request:
# Pull Request

Closes: PRO-1087

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

Added bouncer command to check the number of validator in the active set which witnessed something.
Currently it support 4 different arguments, DOT, BTC, ETH which will make the script wait for the next chaintracking extrinsic being reported for the specified chain and then after a couple of state chain blocks check how many validator witnessed it. (this could take a while for BTC since block production is supposed to be every 10m)

Otherwise HASH can be used followed by the hash to check, this will query the storage item with the given hash and immediately check how many validator reported it, in order to build the correct hash polkaJS can be used as explained in the INSTRUCTION at the beginning of the file.

marcellorigotti commented:
I had to disable the eslint check for restricted syntax otherwise we cannot use `for of`, which unfortunately cannot be replaced with `foreach` when doing async calls.

kylezs commented on bouncer/commands/check_wintesses.ts:
```diff
@@ -0,0 +1,226 @@
+import { blake2AsHex } from '@polkadot/util-crypto';
+import { runWithTimeout, sleep, getChainflipApi } from '../shared/utils';
+
+function hex2bin(hex0x: string) {
```
You can just use `parseInt(someHexString, 16).toString(2)` ? Would have to strip out the 0x first still though

marcellorigotti replied:
This was problematic because of the max integer size in js. Which is way lower than the value represented by the hex, not sure if there is another direct way to convert hex to binary.

kylezs replied:
I see - should at least leave a comment that this is the reason for this function. 

Also, just noticed the file has a typo, should be `check_witnesses`

kylezs commented on bouncer/commands/check_wintesses.ts:
```diff
@@ -0,0 +1,226 @@
+            finalData.new_chain_state,
+          );
+          // obtain the hash of the extrinsic call
+          const blakeHash = blake2AsHex(extrinsic.method.toU8a(), 256);
+          witnessHash.add(blakeHash);
```
Seems like this could be a function - repeated a number of times

kylezs commented on bouncer/commands/check_wintesses.ts:
```diff
@@ -0,0 +1,226 @@
+              const vanity = vanityNames[element].substr(2);
+              const bytes = [];
+              for (let i = 0; i < vanity.length; i += 2) {
+                bytes.push(parseInt(vanity.substr(i, 2), 16));
```
Looks like you actually use the parseInt version here already 🤔

dandanlen commented:
FWIW I think it would be easier / more accurate to add a runtime rpc that takes a call or a hash and looks up the witness votes / does the calculation natively in rust instead of trying to replicate everything in javascript. This would have the further advantage of being easy to use from the command line or polkaJS.

marcellorigotti commented:
EDIT:
- Addded cf_witness_count(hash) as a custom rpc
- Added CLI command count-witnesses <HASH>, which uses the custom rpc to retrieve how many validators are not witnessing something
- Modified the bouncer command to use the newly created custom rpc

dandanlen commented on state-chain/custom-rpc/src/lib.rs:
```diff
@@ -487,6 +489,9 @@ pub trait CustomApi {
 	) -> RpcResult<Option<<cf_chains::Ethereum as Chain>::Transaction>>;
+
+	#[method(name = "witness_count")]
+	fn cf_witness_count(&self, hash: String) -> RpcResult<Option<FailingWitnessValidators>>;
```
I think you can use `state_chain_runtime::Hash` as the type here and it will deserialize correctly. Avoids having to use `from_hex` to convert the string.

dandanlen commented on state-chain/runtime/src/lib.rs:
```diff
@@ -1342,6 +1344,53 @@ impl_runtime_apis! {
+						// we can safely unwrap since it is impossible to retrieve indices greater than the number of elements
+						let validator: &AccountId = validators.iter().nth(elem).unwrap();
+						// we can safely unwrap since it is impossible to not have a vanity record for an authority
+						let vanity: &Vec<u8> = vanity_names.get(validator).unwrap();
```
AFAIK This isn't true - authorities are not required to have a vanity name.

marcellorigotti replied:
You are right, I though even with no vanity name they still had an entry in the map (which turns out is not the case)

dandanlen commented on state-chain/runtime/src/lib.rs:
```diff
@@ -1342,6 +1344,53 @@ impl_runtime_apis! {
 			}
 		}
+
+		fn cf_witness_count(hash: pallet_cf_witnesser::CallHash) -> Option<FailingWitnessValidators> {
```
I think it would be useful to give a full list of who did and didn't vote instead of just the missing votes. Basically a Vec of { id, vanity, vote }.

dandanlen commented on state-chain/runtime/src/lib.rs:
```diff
@@ -1342,6 +1344,53 @@ impl_runtime_apis! {
+
+		fn cf_witness_count(hash: pallet_cf_witnesser::CallHash) -> Option<FailingWitnessValidators> {
+			// helper function to compute the number of bits set and their indices
+			fn count_set_bits_and_indices(bytes: &[u8], size: usize) -> (u32, Vec<usize>) {
```
I think I didn't express very well what I meant with my other comment. 

The witnesser pallet already uses a crate `bitvec` that has utilities to index into a vector of bits. 

You could load the votes into a BitVec and then just iterate this to get a Vec of (authority, vote, vanity_name). Something like this:

```rust
let votes = BitVec::from_vec(Votes::<Runtime>::get(callhash)?);
let vote_count = votes.count_ones();
let missing_vote_count = votes.count_zeros();

let result: Vec<_> = votes.iter()
	// by_vals is needed to convert to true/false bool values.
	.by_vals()
	// authorities are stored in the same order as the votes
	.zip(authorities)
	.map(|(vote, account_id)| (
		account_id,
		vote,
		vanity_names[account_id],
	))
	.collect();
```

This can all be implemented as a public fn inside the witnesser pallet.

marcellorigotti commented:
A quick comment on this.
It would be nice if we can include this PR in the next release such that we can then use the bouncer command for mainnet/perseverance (otherwise it won't work since it uses a new runtime rpc)

dandanlen commented on state-chain/pallets/cf-witnesser/src/lib.rs:
```diff
@@ -479,6 +479,25 @@ impl<T: Config> Pallet<T> {
+	pub fn count_votes(callhash: CallHash) -> Option<Vec<(<T as Chainflip>::ValidatorId, bool)>> {
+		let current_epoch = T::EpochInfo::epoch_index();
+		let votes: BitVec<u8, Msb0> =
+			BitVec::from_vec(Votes::<T>::get(current_epoch, callhash).unwrap_or(vec![]));
+		if votes.is_empty() {
+			return None
+		}
```
nit: I think this is equivalent and more idiomatic:

```suggestion
			BitVec::from_vec(Votes::<T>::get(current_epoch, callhash)?);
```

A lot of these variables can also be inlined.

dandanlen commented on api/lib/src/queries.rs:
```diff
@@ -220,6 +220,21 @@ impl QueryApi {
+
+	pub async fn check_witnesses(
+		&self,
+		_block_hash: Option<state_chain_runtime::Hash>,
```
Might be a good idea to allow querying historical blocks... This can be useful for debugging things in the past. It should be a relatively simple addition in the runtime rpc. LMK, I can do it on Monday.

dandanlen commented on api/bin/chainflip-cli/src/main.rs:
```diff
@@ -285,6 +288,21 @@ async fn pre_update_check(api: QueryApi) -> Result<()> {
+	let result = api.check_witnesses(None, hash).await?;
+	match result {
+		Some(value) => {
+			println!("Number of validator who failed to witness it: {}", value.failing_count);
+			println!("List of validator: {:?}", value.validators);
```
nit:

```suggestion
			println!("Number of authorities who failed to witness it: {}", value.failing_count);
			println!("List of witness votes:\n {:?}", value.validators);
```

dandanlen reviewed (commented):
Good work @marcellorigotti 

Just a final thing to add 😅 I think it would be useful to allow querying the witness votes at historical block hashes.
